designer_v2/lib/features/dashboard/dashboard_state.dart

```diff
@@ -25,33 +26,64 @@ class DashboardState extends Equatable {
   /// Currently authenticated user (used for filtering studies)
   final User currentUser;
 
+  final String query;
+
   /// The currently visible list of studies as by the selected filter
   ///
   /// Wrapped in an [AsyncValue] that mirrors the [studies]' async states,
   /// but resolves to a different subset of studies based on the [studiesFilter]
-  AsyncValue<List<Study>> get visibleStudies {
+  AsyncValue<List<Study>> visibleStudies(Set<String> pinnedStudies, String query) {
     return studies.when(
-      data: (studies) => AsyncValue.data(_filterAndSortStudies(studies)),
+      data: (studies) {
+        List<Study> updatedStudies = studiesFilter.apply(studies: studies, user: currentUser).toList();
+        updatedStudies = filter(studiesToFilter: updatedStudies);
+        updatedStudies = sort(pinnedStudies: pinnedStudies, studiesToSort: updatedStudies);
+        return AsyncValue.data(updatedStudies);
+      },
       error: (error, _) => AsyncValue.error(error, StackTrace.current),
       loading: () => const AsyncValue.loading(),
     );
   }
 
-  List<Study> _filterAndSortStudies(List<Study> studies) {
-    final filteredStudies = studiesFilter.apply(studies: studies, user: currentUser).toList();
-    filteredStudies.sort((study, other) => study.title!.compareTo(other.title!));
+  List<Study> filter({List<Study>? studiesToFilter}) {
+    final filteredStudies = studiesToFilter ?? studies.value!;
+    if (query.isNotEmpty) {
+      return filteredStudies.where((s) => s.title!.toLowerCase().contains(query)).toList();
+    }
     return filteredStudies;
   }
 
+  List<Study> sort({required Set<String> pinnedStudies, List<Study>? studiesToSort}) {
+    final sortedStudies = studiesToSort ?? studies.value!;
+    sortedStudies.sort((study, other) => study.title!.compareTo(other.title!));
+    if (pinnedStudies.isNotEmpty) {
+      // Extract pinned studies and remove them from filteredStudies
+      final List<Study> pinned = [];
+      sortedStudies.removeWhere((study) {
+        if (pinnedStudies.contains(study.id)) {
+          pinned.add(study);
+          return true;
+        }
+        return false;
+      });
+
+      // Insert pinned studies at the beginning of the filteredStudies list
+      sortedStudies.insertAll(0, pinned);
+    }
```

Comment:
Would it be possible to simplify this by just adding the `is pinned?`-condition to the sorting predicate? I.e. if `pinnedStudies.contains(other) - pinnedStudies.contains(study)` is not 0 we return this difference and else the current predicate